File: lambda/send-email/lambda_function.js
const AWS = require('aws-sdk');
const sgMail = require('@sendgrid/mail');

const dynamodb = new AWS.DynamoDB.DocumentClient({ region: 'us-east-1' });
const s3 = new AWS.S3({ region: 'ap-south-2' });

const DYNAMODB_TABLE = process.env.DYNAMODB_TABLE || 'vmail-emails';
const S3_BUCKET = process.env.S3_BUCKET || 'vmail-emails-059409992687';
const SENDGRID_API_KEY = process.env.SENDGRID_API_KEY;
const SENDGRID_FROM_EMAIL = process.env.SENDGRID_FROM_EMAIL || 'noreply@yourdomain.com';

// Initialize SendGrid
if (SENDGRID_API_KEY) {
  sgMail.setApiKey(SENDGRID_API_KEY);
}

exports.handler = async (event, context) => {
  try {
    // Extract user info from Cognito authorizer
    const userId = event.requestContext.authorizer.claims.sub;
    const userEmail = event.requestContext.authorizer.claims.email;

    // Parse request body
    const body = JSON.parse(event.body);
    const { to, subject, body: emailBody, cc = [], bcc = [], attachments = [] } = body;

    // Generate unique email ID
    const emailId = `${userId}-${Date.now()}`;
    const timestamp = new Date().toISOString();

    // Prepare email message for SendGrid
    const msg = {
      to: Array.isArray(to) ? to : [to],
      from: {
        email: SENDGRID_FROM_EMAIL,
        name: userEmail
      },
      replyTo: userEmail,
      subject: subject,
      text: emailBody.replace(/<[^>]*>/g, ''),
      html: emailBody
    };

    // Add CC and BCC if provided
    if (cc && cc.length > 0) {
      msg.cc = cc;
    }
    if (bcc && bcc.length > 0) {
      msg.bcc = bcc;
    }

    // Add attachments if provided
    if (attachments && attachments.length > 0) {
      msg.attachments = attachments.map(att => ({
        content: att.data,
        filename: att.filename,
        type: att.contentType,
        disposition: 'attachment'
      }));
    }

    // Send email using SendGrid
    const sendResult = await sgMail.send(msg);
    const messageId = sendResult[0].headers['x-message-id'] || emailId;

    // Store full email in S3
    const s3Key = `emails/${userId}/${emailId}.json`;
    await s3.putObject({
      Bucket: S3_BUCKET,
      Key: s3Key,
      Body: JSON.stringify({
        subject,
        body: emailBody,
        from: userEmail,
        to: Array.isArray(to) ? to : [to],
        cc,
        bcc,
        attachments,
        timestamp,
        messageId
      }),
      ContentType: 'application/json'
    }).promise();

    // Store metadata in DynamoDB
    await dynamodb.put({
      TableName: DYNAMODB_TABLE,
      Item: {
        emailId,
        userId,
        from: userEmail,
        to: Array.isArray(to) ? to : [to],
        cc,
        bcc,
        subject,
        preview: emailBody.replace(/<[^>]*>/g, '').substring(0, 100),
        timestamp,
        folder: 'sent',
        read: true,
        hasAttachments: attachments.length > 0,
        s3Key,
        messageId
      }
    }).promise();

    return {
      statusCode: 200,
      headers: getCorsHeaders(),
      body: JSON.stringify({
        message: 'Email sent successfully via SendGrid',
        emailId,
        messageId
      })
    };
  } catch (error) {
    console.error('Error sending email:', error);
    return {
      statusCode: 500,
      headers: getCorsHeaders(),
      body: JSON.stringify({
        message: `Error sending email: ${error.message}`
      })
    };
  }
};

function getCorsHeaders() {
  return {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Headers': 'Content-Type,Authorization',
    'Access-Control-Allow-Methods': 'OPTIONS,POST,GET,PUT,DELETE'
  };
}

File: lambda/send-email/lambda_function.py
import json
import boto3
import os
from datetime import datetime
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText
from email.mime.application import MIMEApplication
import base64

# Initialize AWS clients
ses = boto3.client('ses', region_name='us-east-1')
dynamodb = boto3.resource('dynamodb', region_name='us-east-1')
s3 = boto3.client('s3', region_name='ap-south-2')

# Environment variables
DYNAMODB_TABLE = os.environ.get('DYNAMODB_TABLE', 'vmail-emails')
S3_BUCKET = os.environ.get('S3_BUCKET', 'vmail-emails-bucket')

def lambda_handler(event, context):
    """
    Lambda function to send emails via AWS SES
    """
    try:
        # Parse request body
        body = json.loads(event.get('body', '{}'))

        # Extract user info from Cognito authorizer
        user_email = event['requestContext']['authorizer']['claims']['email']
        user_id = event['requestContext']['authorizer']['claims']['sub']

        # Validate required fields
        to_addresses = body.get('to', [])
        if isinstance(to_addresses, str):
            to_addresses = [to_addresses]

        subject = body.get('subject', '')
        email_body = body.get('body', '')
        cc_addresses = body.get('cc', [])
        bcc_addresses = body.get('bcc', [])
        attachments = body.get('attachments', [])

        if not to_addresses or not subject:
            return {
                'statusCode': 400,
                'headers': get_cors_headers(),
                'body': json.dumps({'message': 'Missing required fields: to, subject'})
            }

        # Generate unique email ID
        email_id = f"{user_id}-{int(datetime.now().timestamp() * 1000)}"
        timestamp = datetime.now().isoformat()

        # Create MIME message
        msg = MIMEMultipart()
        msg['Subject'] = subject
        msg['From'] = user_email
        msg['To'] = ', '.join(to_addresses)

        if cc_addresses:
            msg['Cc'] = ', '.join(cc_addresses)

        # Add body
        msg.attach(MIMEText(email_body, 'plain'))

        # Add attachments
        for attachment in attachments:
            try:
                file_data = base64.b64decode(attachment['data'])
                part = MIMEApplication(file_data)
                part.add_header('Content-Disposition', 'attachment',
                              filename=attachment['filename'])
                msg.attach(part)
            except Exception as e:
                print(f"Error attaching file: {str(e)}")

        # Send email via SES
        all_recipients = to_addresses + cc_addresses + bcc_addresses
        ses_response = ses.send_raw_email(
            Source=user_email,
            Destinations=all_recipients,
            RawMessage={'Data': msg.as_string()}
        )

        # Store email body in S3
        s3_key = f"emails/{user_id}/{email_id}.json"
        s3.put_object(
            Bucket=S3_BUCKET,
            Key=s3_key,
            Body=json.dumps({
                'subject': subject,
                'body': email_body,
                'to': to_addresses,
                'cc': cc_addresses,
                'bcc': bcc_addresses,
                'attachments': attachments,
                'timestamp': timestamp
            }),
            ContentType='application/json'
        )

        # Store metadata in DynamoDB
        table = dynamodb.Table(DYNAMODB_TABLE)
        table.put_item(
            Item={
                'emailId': email_id,
                'userId': user_id,
                'from': user_email,
                'to': to_addresses,
                'cc': cc_addresses,
                'bcc': bcc_addresses,
                'subject': subject,
                'preview': email_body[:100],
                'timestamp': timestamp,
                'folder': 'sent',
                'read': True,
                'hasAttachments': len(attachments) > 0,
                's3Key': s3_key,
                'messageId': ses_response['MessageId']
            }
        )

        return {
            'statusCode': 200,
            'headers': get_cors_headers(),
            'body': json.dumps({
                'message': 'Email sent successfully',
                'emailId': email_id,
                'messageId': ses_response['MessageId']
            })
        }

    except Exception as e:
        print(f"Error sending email: {str(e)}")
        return {
            'statusCode': 500,
            'headers': get_cors_headers(),
            'body': json.dumps({
                'message': f'Error sending email: {str(e)}'
            })
        }

def get_cors_headers():
    """Return CORS headers"""
    return {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Headers': 'Content-Type,Authorization',
        'Access-Control-Allow-Methods': 'OPTIONS,POST,GET,PUT,DELETE'
    }
